src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -34,12 +34,26 @@ public PipeCompletionCallbacks TryComplete(Exception exception = null)
                 _isCompleted = true;
                 if (exception != null)
                 {
-                    _exceptionInfo = ExceptionDispatchInfo.Capture(exception);
+                    CaptureExceptionStackTrace(exception);
                 }
             }
+
             return GetCallbacks();
         }
 
+        private void CaptureExceptionStackTrace(Exception exception)
+        {
+            try
+            {
+                // Throw to capture caller stack trace in exception
+                ExceptionDispatchInfo.Capture(exception).Throw();
+            }
+            catch (Exception ex)
+            {
+                _exception = ex;
+            }
```

Comment:
I don't think this is going to do what you want it to do.

Reply:
Hmm... true.

Is there a way of sticking the current stack trace into an exception?

Reply:
> Is there a way of sticking the current stack trace into an exception?

Not that I know of, unfortunately.  It's something we'd benefit from in a variety of places, as it's common in async programming to create exception instances that are then transferred somewhere else rather than thrown.
cc: @jkotas

Reply:
Issue is a lot of the exceptions start at
```
at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
```
Which is the EDI Throw; (because a new Exception is passed in rather than a thrown one); whereas I want it to start before, so we know what creates the exception so it flows.

Can't just throw the exception as it may be a thrown one and that would overwrite its trace.

Reply:
Something like the following would be good
```

```

Reply:
Something like the following would be good
```
ExceptionDispatchInfo ExceptionDispatchInfo.Capture(Exception ex, bool captrueStackTraceIfNotThrown = false)
```

Reply:
> Is there a way of sticking the current stack trace into an exception?

Would storing the EDI instead of Exception solve the problem you are trying to solve? The EDI allows you to reset the exception stacktrace on repeated throws. It works fine as long as the same exception instance is not thrown in parallel on multiple threads.

> It's common in async programming to create exception instances that are then transferred somewhere else rather than thrown.

If you set "stop on first chance exception in the debugger", would you like to stop in the place where the exception instance is created; or only in the new transfer place? The EDI design assumes that you want to stop in the place where the exception is created.

Reply:
@jkotas, the problem is two-fold:
- this exception actually could be reused across multiple threads, which is why this PR is moving away from ExceptionDispatchInfo and instead wrapping the exception
- this exception was never thrown, but rather created from scratch to represent some condition.  It'd be helpful to know the stack trace where the exception was created, but whether it's ExceptionDispatchInfo.Capture'd and later thrown or whether it's wrapped in another exception, that original stack trace where the exception was created is lost, as there's no way to create an exception and see "store the current stack trace".

Reply:
I am not following. The way to store the current stacktrace as the exception stack trace is to throw and catch it. What would be the test that demonstrates the difference between new Exception+throw+catch and this new way to store the current stacktrace into the exception?

Reply:
> The way to store the current stacktrace as the exception stack trace is to throw and catch it.

If you throw and catch it in the same frame, all you get in the stack trace is that frame, not the full stack.

Consider something like:
https://github.com/dotnet/coreclr/blob/87f4eff9188456c26ffab6efbcc725110630a3dd/src/System.Private.CoreLib/shared/System/IO/FileStreamCompletionSource.Win32.cs#L193
We're creating an exception based on something that's occurred, and storing it into a Task.  The current stack trace that includes CompleteCallback will not be anywhere in that exception when it's eventually propagated to the caller awaiting that Task somewhere else.

Reply:
> What would be the test that demonstrates the difference between new Exception+throw+catch and this new way to store the current stacktrace into the exception?

```C#
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main() => A();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void A() => B();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void B() => C();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static void C()
    {
        Console.WriteLine("Full stack:");
        Console.WriteLine(Environment.StackTrace);
        Console.WriteLine();

        Console.WriteLine("Exception throw+catch stack:");
        try
        {
            throw new Exception();
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }
}
```
outputs:
```
Full stack:
   at System.Environment.get_StackTrace()
   at Program.C() in C:\Users\stoub\source\repos\ConsoleApp18\ConsoleApp18\Program.cs:line 18
   at Program.B() in C:\Users\stoub\source\repos\ConsoleApp18\ConsoleApp18\Program.cs:line 12
   at Program.A() in C:\Users\stoub\source\repos\ConsoleApp18\ConsoleApp18\Program.cs:line 9
   at Program.Main() in C:\Users\stoub\source\repos\ConsoleApp18\ConsoleApp18\Program.cs:line 6

Exception throw+catch stack:
   at Program.C() in C:\Users\stoub\source\repos\ConsoleApp18\ConsoleApp18\Program.cs:line 24
```

Reply:
Got it. Thanks

Reply:
throw+catch stores the stack from the point of the throw to the catch. 

However if you are using an unthrown exception; say when completing a TaskCompletionSource; Task etc; while it may be "active" and have a stacktrace (when all is good, and Edi works for gluing); if it was just passed as a `new Exception` then it has no stacktrace and there is no source for the exception so it initiates from the secondary throw point.

Its the second scenario of a non-Active exception that I'd want to capture the current stack to; so it can have a source point to trace the flow; rather than just be unrooted.

Reply:
Example use case:
```csharp
public class MyClass
{
    private TaskCompletionSource<object> _tcs = new TaskCompletionSource<object>();

    public Task Task => _tcs.Task;

    public void FailNoStack()
    {
        _tcs.SetException(new Exception("Failed"));
    }

    public void FailMildStack()
    {
        try
        {
            throw new Exception("Failed");
        }
        catch (Exception ex)
        {
            _tcs.SetException(ex);
        }
    }
}
```
`FailNoStack` seems preferable in usage, but gives no source suggesting the exception originates in the `TaskCompletionSource`.

`FailMildStack` gives some stack; but is very heavyweight and doesn't seem a good pattern to use just to get some stack prior to the TCS/Task.Result throw

Reply:
> FailMildStack gives some stack

Not really: all it gives you is `FailMildStack` itself.

Reply:
Mild(ish) :)

Second part of this [from the ExceptionDispatchInfo.Capture docs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.exceptionservices.exceptiondispatchinfo.capture?view=netcore-3.0#remarks) is the issue 

> If the exception is active when it is captured, the current stack trace information and Watson information that is contained in the exception is stored. 
>
> **If it is inactive, that is, if it has not been thrown, it will not have any stack trace information or Watson information.**